simulation/chart.py

```diff
@@ -1,232 +1,438 @@
-        chart.setYRange(*y_range)
-
-    layout.nextRow()
-
-    return chart
-
-
-def add_line_plot(
-    chart: pg.PlotItem, name: str, color=None, width: Optional[float] = None, style=None
-) -> pg.PlotDataItem:
-    """Adds a line plot to the chart.
+    R: int
+    G: int
+    B: int
+    A: Optional[int]
 
-    Args:
-        name (str): Name (legend title) of the plot
-        color (Any, optional): Line color of the plot. Defaults to None.
-        width (float, optional): Width of the plot line. Defaults to None.
-        style (Any, optional): Style of the plot line. Defaults to None.
+
+# Single-character string representing a predefined color
+NamedColor = Literal[
+    "r",  # red
+    "g",  # green
+    "b",  # blue
+    "c",  # cyan
+    "m",  # magenta
+    "y",  # yellow
+    "k",  # key (black)
+    "w",  # white
+]
+
+
+class IndexedColor(NamedTuple):
+    """Color from a single index. Useful for stepping through a predefined list of colors.
+
+    Attributes:
+        index (int): predefined color index
+        hues (int): number of steps to generate
     """
-    config = {}
+
+    index: int
+    hues: int
+
+
+Color = Union[RGBAColor, NamedColor, IndexedColor, int, float, str]
+"""Color based on model, predefined values, single greyscale value (0.0 - 1.0) , color index or string representing color (“#RGB”,“#RGBA”,“#RRGGBB”,“#RRGGBBAA”)."""
+
+
+@multimethod
+def _make_color(color: RGBAColor):  # pyright:ignore [reportGeneralTypeIssues]
+    """Convert RGBA color value to QColor"""
+    if color.A is None:
+        return pg.mkColor((color.R, color.G, color.B))
+    else:
+        return pg.mkColor((color.R, color.G, color.B, color.A))
+
+
+@multimethod
+def _make_color(color: IndexedColor):  # pyright:ignore [reportGeneralTypeIssues]
+    """Convert Indexed color value to QColor"""
+    return pg.mkColor((color.index, color.hues))
+
+
+@multimethod
+def _make_color(color: Union[int, float, str]):
+    return pg.mkColor(color)
+
+
+class PenStyle(Enum):
+    """Pen style for drawing lines"""
+
+    # NoPen                    = QtCore.Qt.PenStyle.NoPen
+    SolidLine = QtCore.Qt.PenStyle.SolidLine  # pyright:ignore [reportGeneralTypeIssues]
+    DashLine = QtCore.Qt.PenStyle.DashLine  # pyright:ignore [reportGeneralTypeIssues]
+    DotLine = QtCore.Qt.PenStyle.DotLine  # pyright:ignore [reportGeneralTypeIssues]
+    DashDotLine = (
+        QtCore.Qt.PenStyle.DashDotLine  # pyright:ignore [reportGeneralTypeIssues]
+    )
+    DashDotDotLine = (
+        QtCore.Qt.PenStyle.DashDotDotLine  # pyright:ignore [reportGeneralTypeIssues]
+    )
+    # CustomDashLine           = QtCore.Qt.PenStyle.CustomDashLine
+    # MPenStyle                = QtCore.Qt.PenStyle.MPenStyle
+
+
+def _make_pen(
+    color: Optional[Color] = None,
+    width: Optional[float] = None,
+    style: Optional[PenStyle] = None,
+):
+    """Create a QPen from provided parameters"""
+    config: Dict[str, Any] = {}
+
     if color is not None:
```

Comment:
Instead of doing this, you can dispatch to a `_make_color` function that dispatches on `None`. Gets rid of some explicit RTTI.

Reply:
Can you clarify it a little bit more, do you expect me to write an overload with None type argument ?

Reply:
````python
@singledispatch
def _make_color():
    return None
````

Reply:
You'd use `None` as the type of one of the function methods.

Reply:
https://github.com/semiotic-ai/autoagora-agents/pull/21/commits/d6582cf6a8d6916239e09a4f62d266c848672a60